### ISSUE TITLE
Detailed questions about CausalVAE intervention

### ISSUE THREAD

**Issue by EternityZY** [huawei-noah/trustworthyAI#59]
Hello, how did CausalVAE intervention operation? In the published source code, intervention is to forcibly assign a concept to a fixed integer by using "mask_z", as follows:
```
z_mask = torch.ones(q_m.size()[0], self.z1_dim, self.z2_dim).to(device) * adj       
decode_m[:, mask, :] = z_mask[:, mask, :]                       
decode_v[:, mask, :] = z_mask[:, mask, :]    
```
In the test result, the result is very poor. Where did I make a mistake?

The original input image is this
![true_0](https://user-images.githubusercontent.com/25786054/162562132-f8e36d2c-6819-4c60-97c2-40476fb28ec9.png)

Output image with intervention of 1 for the 0th concept
![reconstructed_image_0_0](https://user-images.githubusercontent.com/25786054/162562119-6c0bcb88-0e7e-4ba9-8c48-d8ea2126e2c3.png)

Should I modify the published code to achieve the effect in the paper?

**Comment by shaido987** [comment]
Hello, did you make any changes to the source code to get this result? How did you run the code?

**Comment by EternityZY** [comment]
> Hello, did you make any changes to the source code to get this result? How did you run the code?

Thank you for your help. I didn't modify the source code and ran python ./run_flow.py directly. After the training, run python ./run_pendulum.py

**Comment by shaido987** [comment]
I see, I assume the synthetic data was generated correctly as well (with `./causal_data/pendulum.py`)? Flow and pendulum are two different datasets so they shouldn't affect each other. 

Did you run the intervention using `python ./inference_pendulum.py` after the training finished?

**Comment by EternityZY** [comment]
> I see, I assume the synthetic data was generated correctly as well (with `./causal_data/pendulum.py`)? Flow and pendulum are two different datasets so they shouldn't affect each other.
> 
> Did you run the intervention using `python ./inference_pendulum.py` after the training finished?

I'm very sorry for the mistake,
ran python ./run_pendulum.py directly. After the training, run python ./inference_pendulum.py

**Comment by shaido987** [comment]
Happy to help, hopefully this should solve it for you.

**Comment by EternityZY** [comment]
> Happy to help, hopefully this should solve it for you.

You misunderstood me. What I just wrote was wrong. I ran run_pendulum.py and then inference_pendulum.py, and the problem raised on the page appeared.
I didn't modify the code, and the running sequence is in accordance with readme.md. I found that the effect is not ideal in the published source code. I would like to ask which part of the source code should be corresponding to the intervention operation in this paper?

**Comment by shaido987** [comment]
I see, I've been in contact with one of the authors of the paper to check this issue.

**Comment by ymy4323460** [comment]
> Hello, how did CausalVAE intervention operation? In the published source code, intervention is to forcibly assign a concept to a fixed integer by using "mask_z", as follows:
> 
> ```
> z_mask = torch.ones(q_m.size()[0], self.z1_dim, self.z2_dim).to(device) * adj       
> decode_m[:, mask, :] = z_mask[:, mask, :]                       
> decode_v[:, mask, :] = z_mask[:, mask, :]    
> ```
> 
> In the test result, the result is very poor. Where did I make a mistake?
> 
> The original input image is this ![true_0](https://user-images.githubusercontent.com/25786054/162562132-f8e36d2c-6819-4c60-97c2-40476fb28ec9.png)
> 
> Output image with intervention of 1 for the 0th concept ![reconstructed_image_0_0](https://user-images.githubusercontent.com/25786054/162562119-6c0bcb88-0e7e-4ba9-8c48-d8ea2126e2c3.png)
> 
> Should I modify the published code to achieve the effect in the paper?


After running './run_pendulum.py' you can run './inference_pendulum.py' to get intervention results where all the concept values are fixed as 0. The problem can be solved by following two steps:

1. Modifying 0 in line 61 'ut.load_model_by_name(lvae, 0)', inference_xxx.py  as the biggest epoch number of saved models in './checkpoint'. For example, if you save the model optimized at epoch 70, set 'ut.load_model_by_name(lvae, 70)'.

2. If the first solution doesn't work. Try to additionally use the annotated code on lines 251-252, './codebase/models/nns/mask.py'.
The learning process of the causal graph in CausalVAE actually needs pre-train (augmented Lagrangian multiplier method). Details could be found in the Appendix part of the paper. The initialization of causal graph A is really important. We do not present related code in this repo, you can directly try to init A as the upper triangular matrix or the adjacency matrix of the true graph.
